src/controllers/rendering/controllerrenderingengine.cpp

```diff
@@ -0,0 +1,436 @@
+            if (m_pControllerEngine) {
+                m_pControllerEngine->resume();
+            }
+
+            return;
+        };
+    }
+
+    if (m_pControllerEngine) {
+        m_pControllerEngine->resume();
+    }
+    QImage fboImage(m_screenInfo.size, m_screenInfo.pixelFormat);
+
+    VERIFY_OR_DEBUG_ASSERT(m_fbo->bind()) {
+        kLogger.warning() << "Couldn't bind the FBO.";
+    }
+    GLenum glError;
+    m_context->functions()->glFlush();
+    glError = m_context->functions()->glGetError();
+    VERIFY_OR_DEBUG_ASSERT(glError == GL_NO_ERROR) {
+        kLogger.warning() << "GLError: " << glError;
+        lock.unlock();
+        finish();
+        return;
+    }
+    if (m_screenInfo.endian != std::endian::native) {
+        m_context->functions()->glPixelStorei(GL_PACK_SWAP_BYTES, GL_TRUE);
+    }
+    glError = m_context->functions()->glGetError();
+    VERIFY_OR_DEBUG_ASSERT(glError == GL_NO_ERROR) {
+        kLogger.warning() << "GLError: " << glError;
+        lock.unlock();
+        finish();
+        return;
+    }
+
+    QDateTime timestamp = QDateTime::currentDateTime();
+    m_renderControl->render();
+    m_renderControl->endFrame();
+
+    // Flush any remaining GL errors
+    while (m_context->functions()->glGetError()) {
+    }
+    {
+        ScopedTimer t(u"ControllerRenderingEngine::renderFrame::glReadPixels");
+        m_context->functions()->glReadPixels(0,
+                0,
+                m_screenInfo.size.width(),
+                m_screenInfo.size.height(),
+                m_GLDataFormat,
+                m_GLDataType,
+                fboImage.bits());
+    }
+    glError = m_context->functions()->glGetError();
+    VERIFY_OR_DEBUG_ASSERT(glError == GL_NO_ERROR) {
+        kLogger.warning() << "GLError: " << glError;
+        lock.unlock();
+        finish();
+        return;
+    }
+    VERIFY_OR_DEBUG_ASSERT(!fboImage.isNull()) {
+        kLogger.warning() << "Screen frame is null!";
+    }
+    VERIFY_OR_DEBUG_ASSERT(m_fbo->release()) {
+        kLogger.debug() << "Couldn't release the FBO.";
+    }
+
+    fboImage.mirror(false, true);
+
+    emit frameRendered(m_screenInfo, fboImage, timestamp);
+
+    m_context->doneCurrent();
+}
+
+bool ControllerRenderingEngine::stop() {
+    emit stopRequested();
+    return m_pThread->wait();
+}
+
+void ControllerRenderingEngine::send(Controller* controller, const QByteArray& frame) {
+    ScopedTimer t(u"ControllerRenderingEngine::send");
+    if (!frame.isEmpty()) {
+        controller->sendBytes(frame);
+    }
+
+    if (CmdlineArgs::Instance()
+                    .getControllerDebug()) {
+        auto endOfRender = Clock::now();
+        kLogger.debug()
+                << "Frame took "
+                << std::chrono::duration_cast<std::chrono::milliseconds>(
+                           endOfRender - m_nextFrameStart)
+                           .count()
+                << "milliseconds and frame has" << frame.size() << "bytes";
+    }
```

Comment:
No because the time a frame takes must include the time it takes to also transform and send it to the device to be valuable. For example on the S4 Mk3, it takes < 1ms to render,  ~2ms to transform, ~80ms to send a full frame, so it wouldn't be useful to a controller developer to say the `Frame took 1ms` as to their eyes, it should include the whole cycle of rendering all the way to the device.

Reply:
ah right. that makes sense. I got confused by the name `endOfRender`.

Reply:
Let me rename that to `endOfFrameCycle`